mlir/include/mlir/Dialect/LLVMIR/LLVMIntrinsicOps.td

```diff
@@ -311,6 +311,47 @@ def LLVM_InvariantEndOp : LLVM_ZeroResultIntrOp<"invariant.end", [2],
       "qualified(type($ptr))";
 }
 
+// Constrained Floating-Point Intrinsics
+
+class LLVM_ConstrainedIntr<string mnem, int numArgs, bit hasRoundingMode>
+    : LLVM_OneResultIntrOp<"experimental.constrained." # mnem,
+                           /*overloadedResults=*/[0],
+                           /*overloadedOperands=*/[0],
+                           /*traits=*/[Pure, DeclareOpInterfaceMethods<ExceptionBehaviorOpInterface>]
+                           # !cond(
+                               !gt(hasRoundingMode, 0) : [DeclareOpInterfaceMethods<RoundingModeOpInterface>],
+                               true : []),
+                           /*requiresFastmath=*/0,
+                           /*immArgPositions=*/[],
+                           /*immArgAttrNames=*/[]> {
+  dag regularArgs = !dag(ins, !listsplat(LLVM_Type, numArgs), !foreach(i, !range(numArgs), "arg_" #i));
+  dag attrArgs = !con(!cond(!gt(hasRoundingMode, 0) : (ins ValidRoundingModeAttr:$roundingmode),
```

Comment:
If the rounding mode attribute is always present (as per your last comment), do we still need this conditional here?

Reply:
It's always present if the operation implements the `RoundingModeOpInterface` interface. My last patch simplifies the path in which that's true. We'll have constrained operations with no rounding mode, e.g., `fpext`. My patch makes adding a new intrinsic of that kind as simple as:
```
def LLVM_ConstrainedFPNameIntr
    : LLVM_ConstrainedIntr<"name", /*numArgs=*/n, /*hasRoundingMode=*/b> {
  let assemblyFormat = ...;
}
```
This is ok as is